lib/api/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+pub trait CompiledModule {
+    fn new(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_binary(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_binary_unchecked(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module>;
+    fn from_file(file: impl AsRef<std::path::Path>) -> error::CompileResult<Module>;
+
+    fn validate(bytes: impl AsRef<[u8]>) -> error::CompileResult<()>;
+}
+
+use wasmer_runtime_core::backend::Compiler;
+
+/// Copied from runtime core; TODO: figure out what we want to do here
+pub fn default_compiler() -> impl Compiler {
+    #[cfg(any(
+        all(
+            feature = "default-backend-llvm",
+            not(feature = "docs"),
+            any(
+                feature = "default-backend-cranelift",
+                feature = "default-backend-singlepass"
+            )
+        ),
+        all(
+            not(feature = "docs"),
+            feature = "default-backend-cranelift",
+            feature = "default-backend-singlepass"
+        )
+    ))]
+    compile_error!(
+        "The `default-backend-X` features are mutually exclusive.  Please choose just one"
+    );
+
+    #[cfg(all(feature = "default-backend-llvm", not(feature = "docs")))]
+    use wasmer_llvm_backend::LLVMCompiler as DefaultCompiler;
+
+    #[cfg(all(feature = "default-backend-singlepass", not(feature = "docs")))]
+    use wasmer_singlepass_backend::SinglePassCompiler as DefaultCompiler;
+
+    #[cfg(any(feature = "default-backend-cranelift", feature = "docs"))]
+    use wasmer_clif_backend::CraneliftCompiler as DefaultCompiler;
+
+    DefaultCompiler::new()
+}
+
+// this implementation should be moved
+impl CompiledModule for Module {
+    fn new(bytes: impl AsRef<[u8]>) -> error::CompileResult<Module> {
+        let bytes = bytes.as_ref();
+        wasmer_runtime_core::compile_with(bytes, &default_compiler())
+    }
+
+    fn from_binary(_bytes: impl AsRef<[u8]>) -> error::CompileResult<Module> {
+        todo!("from_binary: how is this different from `new`?")
+    }
+    fn from_binary_unchecked(_bytes: impl AsRef<[u8]>) -> error::CompileResult<Module> {
+        todo!("from_binary_unchecked")
+    }
+    fn from_file(_file: impl AsRef<std::path::Path>) -> error::CompileResult<Module> {
+        todo!("from_file");
+        /*
+        use std::fs;
+        use std::io::Read;
+        let path = file.as_ref();
+        let mut f =
+            fs::File::open(path).map_err(|_| todo!("Current error enum can't handle this case"))?;
+        // TODO: ideally we can support a streaming compilation API and not have to read in the entire file
+        let mut bytes = vec![];
+        f.read_to_end(&mut bytes)
+            .map_err(|_| todo!("Current error enum can't handle this case"))?;
+
+        Module::from_binary(bytes.as_slice())
+        */
+    }
+
+    fn validate(_bytes: impl AsRef<[u8]>) -> error::CompileResult<()> {
+        todo!("validate")
+    }
+}
```

Comment:
Note: Can we have a trait for `Instance` as well?

It should have the fields `module` (returns a reference to the Module), `exports` (the instantiated exports).

Reply:
> It should have the fields module

traits cant' have fields but I can make methods that do that.  Reiterating I'm not a fan of duplicating the names when we'll have to export both -- I think that might end up forcing users to write code like `<wasmer::Module as wasmer::traits::Module>::method`

Reply:
I think the traits are out of scope for the current PR, as they won't live in the `api` crate.  I thought I commented that "traits can't have fields" before but I don't see it, so I'll say it again:

Traits can't have fields but we can have trait methods which do those things.